{
 "cells": [
  {
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2024-11-23T22:48:49.880647Z",
     "start_time": "2024-11-23T22:48:44.993858Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import sqlite3\n",
    "from datetime import datetime\n",
    "import unittest\n",
    "from nltk import app\n",
    "from io import BytesIO\n",
    "\n",
    "from sqlalchemy.testing import db\n",
    "\n",
    "# Run the Web App Routing notebook\n",
    "%run WebAppRouting.ipynb import app\n",
    "\n",
    "# Test Setup, Routing, Test User and Database\n",
    "class FlaskAppTests(unittest.TestCase):\n",
    "\n",
    "    def setUp(self):\n",
    "        self.client = app.test_client()\n",
    "\n",
    "    # Function to create user record in the database\n",
    "    def create_user_record(self, created_at, email, first_name, hashed_password, last_name, username):\n",
    "        # Insert user data into the database\n",
    "        try:\n",
    "            conn = sqlite3.connect('users.db')\n",
    "            c = conn.cursor()\n",
    "            c.execute('''INSERT INTO users (first_name, last_name, username, email, password, created_at)\n",
    "                             VALUES (?, ?, ?, ?, ?, ?)''',\n",
    "                      (first_name, last_name, username, email, hashed_password, created_at))\n",
    "            conn.commit()\n",
    "            conn.close()\n",
    "            print(\"User data inserted into the database\")  # Debugging line\n",
    "\n",
    "        except sqlite3.IntegrityError as e:\n",
    "            raise e\n",
    "    # Gets the user record by username\n",
    "    def get_user_record(self, username):\n",
    "        conn = sqlite3.connect('users.db')\n",
    "        cursor = conn.cursor()\n",
    "        cursor.execute(\"SELECT password, id FROM users WHERE username=?\", (username,))\n",
    "        matches = cursor.fetchone()\n",
    "        conn.close()\n",
    "        return matches\n",
    "\n",
    "\n",
    "###UNIT TESTS###\n",
    "\n",
    "    # TEST 1: Testing the About Page\n",
    "    def test_about(self):\n",
    "        response = self.client.get('/about')\n",
    "        self.assertEqual(response.status_code, 200)\n",
    "        self.assertIn(b'This dataset is taken from Kaggle.com', response.data)\n",
    "\n",
    "    # TEST 2: Testing the Add Patient Page\n",
    "    def test_add_patient(self):\n",
    "        response = self.client.get('/add_patient')\n",
    "        self.assertEqual(response.status_code, 200)\n",
    "        self.assertIn(b'Add a New Patient', response.data)\n",
    "\n",
    "    # TEST 3: Testing the Confirm Delete Page\n",
    "    def test_confirm_delete_patient(self):\n",
    "        patient_id = 45805\n",
    "        response = self.client.post(\"/confirm_delete_patient\", data={'id':str(patient_id)})\n",
    "        self.assertEqual(response.status_code, 200)\n",
    "        self.assertIn(b'Confirm Delete', response.data)\n",
    "        self.assertIn(b'patient with ID', response.data)\n",
    "\n",
    "    # TEST 4: Testing the Contact Page\n",
    "    def test_contact(self):\n",
    "        response = self.client.get('/contact')\n",
    "        self.assertEqual(response.status_code, 200)\n",
    "        self.assertIn(b'We hope you find this application useful.', response.data)\n",
    "\n",
    "    # TEST 5: Testing for the Edit Patient page\n",
    "    def test_edit_patient(self):\n",
    "        patient_id = 45805\n",
    "        data = {\n",
    "            'id': str(patient_id),\n",
    "            'gender': 'Male',\n",
    "            'age': 45,\n",
    "            'hypertension': 0,\n",
    "            'heart_disease': 0,\n",
    "            'ever_married': 'Yes',\n",
    "            'work_type': 'Private Sector',\n",
    "            'residence_type': 'Urban',\n",
    "            'avg_glucose_level': 85.5,\n",
    "            'bmi': 24.1,\n",
    "            'smoking_status': 'Never Smoked',\n",
    "            'stroke': 0\n",
    "        }\n",
    "        response = self.client.post(\"/edit_patient\", data=data, follow_redirects=True)\n",
    "        self.assertEqual(response.status_code, 200)\n",
    "        self.assertIn(b'Patient with ID 45805 has been updated.', response.data)\n",
    "        self.assertIn(b'<form method=\"GET\" action=\"/search\">', response.data)\n",
    "\n",
    "    # TEST 6: Testing the Home Page\n",
    "    def test_home(self):\n",
    "        response = self.client.get('/')\n",
    "        self.assertEqual(response.status_code, 200)\n",
    "        self.assertIn(b'This dataset is for use by Miskatonic staff to help improve patient outcomes through identifying potential factors that may lead to a stroke in the future.', response.data)\n",
    "\n",
    "    # TEST 7: Testing the Index page (Rendering)\n",
    "    def test_mongodb_data_page(self):\n",
    "        response = self.client.get('/data')\n",
    "        self.assertEqual(response.status_code, 200)\n",
    "        self.assertIn(b'<title>MongoDB Data</title>', response.data)\n",
    "        self.assertIn(b'<h1>MongoDB Collections Data</h1>', response.data)\n",
    "\n",
    "    # TEST 8: Testing the Index Page Functionality\n",
    "    def test_mongodb_data_functionality(self):\n",
    "        response = self.client.get('/data')\n",
    "        self.assertEqual(response.status_code, 200)\n",
    "        self.assertIn(b'<title>MongoDB Data</title>', response.data)\n",
    "        self.assertIn(b'<h1>MongoDB Collections Data</h1>', response.data)\n",
    "\n",
    "        headers = [\n",
    "            b'<th>id</th>',\n",
    "            b'<th>gender</th>',\n",
    "            b'<th>age</th>',\n",
    "            b'<th>hypertension</th>',\n",
    "            b'<th>heart_disease</th>',\n",
    "            b'<th>ever_married</th>',\n",
    "            b'<th>work_type</th>',\n",
    "            b'<th>Residence_type</th>',\n",
    "            b'<th>avg_glucose_level</th>',\n",
    "            b'<th>bmi</th>',\n",
    "            b'<th>smoking_status</th>',\n",
    "            b'<th>stroke</th>'\n",
    "        ]\n",
    "\n",
    "        for header in headers:\n",
    "            self.assertIn(header, response.data)\n",
    "\n",
    "    # TEST 9: Testing the Logged In Page\n",
    "    def test_logged_in(self):\n",
    "        response = self.client.get('/logged_in')\n",
    "        self.assertEqual(response.status_code, 200)\n",
    "        self.assertIn(b'You are Now Logged In', response.data)\n",
    "\n",
    "    # TEST 10: Testing the Login Page (Render)\n",
    "    def test_login_page(self):\n",
    "        response = self.client.get('/login')\n",
    "        self.assertEqual(response.status_code, 200)\n",
    "        self.assertIn(b'<title>Login</title>', response.data)\n",
    "        self.assertIn(b'<h1>Login</h1>', response.data)\n",
    "        self.assertIn(b'<form action=\"/login\" method=\"post\">', response.data)\n",
    "        self.assertIn(b'<label for=\"username\">User Name:</label>', response.data)\n",
    "        self.assertIn(b'id=\"username\"', response.data)\n",
    "        self.assertIn(b'name=\"username\"', response.data)\n",
    "        self.assertIn(b'type=\"text\"', response.data)\n",
    "        self.assertIn(b'required', response.data)\n",
    "        self.assertIn(b'<label for=\"password\">Password:</label>', response.data)\n",
    "        self.assertIn(b'id=\"password\"', response.data)\n",
    "        self.assertIn(b'name=\"password\"', response.data)\n",
    "        self.assertIn(b'type=\"password\"', response.data)\n",
    "        self.assertIn(b'required', response.data)\n",
    "        self.assertIn(b'<input type=\"submit\" value=\"Login\">', response.data)\n",
    "        self.assertIn(b'<a href=\"/\">Return To The Homepage</a>', response.data)\n",
    "\n",
    "    # TEST 11: Testing the Login Page Functionality\n",
    "    def test_login_functionality(self):\n",
    "        with self.client as client:\n",
    "            response = client.post('/login', data=dict(\n",
    "                username='testuser',\n",
    "                password='testpass'\n",
    "            ), follow_redirects=True)\n",
    "            self.assertEqual(response.status_code, 200)\n",
    "            self.assertIn(b'Login', response.data)\n",
    "\n",
    "\n",
    "    # TEST 12: Testing the Register Page\n",
    "\n",
    "    # TEST 13: Testing the Success Page\n",
    "    def test_success(self):\n",
    "        first_name = \"John\" # Example first name\n",
    "        response = self.client.get(f'/success/{first_name}', follow_redirects=True)\n",
    "        self.assertEqual(response.status_code, 200)\n",
    "        self.assertIn(b'Success! John you are now registered!', response.data)\n",
    "\n",
    "    # TEST 14: Testing the Upload Page (Rendering)\n",
    "    def test_upload_csv_page(self):\n",
    "        response = self.client.get('/upload')\n",
    "        self.assertEqual(response.status_code, 200)\n",
    "        self.assertIn(b'<title>Upload CSV File</title>', response.data)\n",
    "        self.assertIn(b'<h2>Upload CSV File</h2>', response.data)\n",
    "        self.assertIn(b'<form method=\"POST\" enctype=\"multipart/form-data\" action=\"/upload\">', response.data)\n",
    "        self.assertIn(b'<input type=\"file\" name=\"file\" required>', response.data)\n",
    "        self.assertIn(b'<button type=\"submit\">Upload</button>', response.data)\n",
    "\n",
    "    # TEST 15: Testing the Upload functionality on the page\n",
    "    def test_upload_csv_file(self):\n",
    "        with self.client as client:\n",
    "            data = {\n",
    "                'file': (BytesIO(b'id,gender,age,hypertension,heart_disease,ever_married,work_type,Residence_type,avg_glucose_level,bmi,smoking_status,stroke\\n1,Male,67,0,1,Yes,Private,Urban,228.69,36.6,formerly smoked,1'), 'stroke_data.csv')\n",
    "            }\n",
    "            response = client.post('/upload', content_type='multipart/form-data', data=data, follow_redirects=True)\n",
    "            self.assertEqual(response.status_code, 200)\n",
    "            self.assertIn(b'Data has been successfully stored!', response.data)\n",
    "\n",
    "            # Check if data was inserted into MongoDB\n",
    "            client = MongoClient('mongodb+srv://0602750:strokedataassignment@strokedata.2lpyv.mongodb.net/')\n",
    "            db = client['HealthcareDataset']\n",
    "            collection = db['stroke_data']\n",
    "            inserted_data = collection.find_one({'id': 1})\n",
    "            self.assertIsNotNone(inserted_data)\n",
    "            self.assertEqual(inserted_data['gender'], 'Male')\n",
    "\n",
    "\n",
    "    # TEST 16: Testing the User Delete Page (Rendering)\n",
    "    def test_user_delete_render(self):\n",
    "        created_at = datetime.now().strftime('%Y-%m-%d %H:%M:%S')\n",
    "        self.create_user_record(created_at, \"john@test.com\", \"John\", \"Password123\", \"Doe\", \"john_doe\")\n",
    "        id = self.get_user_record(\"john_doe\")[1]\n",
    "        with self.client as client:\n",
    "            response = self.client.get(f'/confirm_user_delete/{id}', follow_redirects=True)\n",
    "            self.assertEqual(response.status_code, 200)\n",
    "            self.assertIn(b'Are you sure you want to delete the following user?', response.data)\n",
    "\n",
    "     # TEST 17: Testing the User Delete Page Functionality\n",
    "\n",
    "    # TEST 18: Testing the User Edit Page (Rendering)\n",
    "\n",
    "    # TEST 19: Testing the User Edit Page Functionality\n",
    "\n",
    "    # TEST 20: Testing the Users Page\n",
    "\n",
    "    # TEST 21: Testing the View Patient Page\n",
    "    def test_view_patients(self):\n",
    "        response = self.client.get('/view_patients')\n",
    "        self.assertEqual(response.status_code, 200)\n",
    "        self.assertIn(b'Stroke Dataset', response.data)\n",
    "        # Check for table headers\n",
    "        headers = [\n",
    "            b'<th>ID</th>',\n",
    "            b'<th>Gender</th>',\n",
    "            b'<th>Age</th>',\n",
    "            b'<th>Hypertension</th>',\n",
    "            b'<th>Heart Disease</th>',\n",
    "            b'<th>Ever Married</th>',\n",
    "            b'<th>Work Type</th>',\n",
    "            b'<th>Residence Type</th>',\n",
    "            b'<th>Avg Glucose Level</th>',\n",
    "            b'<th>BMI</th>',\n",
    "            b'<th>Smoking Status</th>',\n",
    "            b'<th>Stroke</th>',\n",
    "            b'<th>Actions</th>'\n",
    "        ]\n",
    "        for header in headers:\n",
    "            self.assertIn(header, response.data)\n",
    "\n",
    "    # TEST 22: Testing the View Patients Page with Search Query\n",
    "    def test_view_patients_search(self):\n",
    "        search_query = \"45805\"\n",
    "        response = self.client.get(f'/view_patients?search={search_query}')\n",
    "        self.assertEqual(response.status_code, 200)\n",
    "        self.assertIn(search_query.encode(), response.data)\n",
    "        self.assertIn(b'<title>Stroke Dataset</title>', response.data)\n",
    "        self.assertIn(b'<h1>Stroke Dataset</h1>', response.data)\n",
    "        self.assertIn(b'<input type=\"text\" name=\"query\" placeholder=\"Search\" value=\"45805\">', response.data)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "unittest.main(argv=[''], verbosity=2, exit=False)\n"
   ],
   "id": "initial_id",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * Serving Flask app 'WebAppRouting'\n",
      " * Debug mode: off\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.\n",
      " * Running on http://127.0.0.1:5000\n",
      "Press CTRL+C to quit\n",
      "test_about (__main__.FlaskAppTests.test_about) ... ok\n",
      "test_add_patient (__main__.FlaskAppTests.test_add_patient) ... ok\n",
      "test_confirm_delete_patient (__main__.FlaskAppTests.test_confirm_delete_patient) ... ok\n",
      "test_contact (__main__.FlaskAppTests.test_contact) ... ok\n",
      "test_edit_patient (__main__.FlaskAppTests.test_edit_patient) ... C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py:865: ResourceWarning: Unclosed MongoClient opened at:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 701, in start\n",
      "    self.io_loop.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\windows_events.py\", line 322, in run_forever\n",
      "    super().run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 641, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 1987, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 523, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 429, in dispatch_shell\n",
      "    await result\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 767, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 429, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3267539545.py\", line 261, in <module>\n",
      "    unittest.main(argv=[''], verbosity=2, exit=False)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 105, in __init__\n",
      "    self.runTests()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 281, in runTests\n",
      "    self.result = testRunner.run(self.test)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\runner.py\", line 240, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 690, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 634, in run\n",
      "    self._callTestMethod(testMethod)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 589, in _callTestMethod\n",
      "    if method() is not None:\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3267539545.py\", line 88, in test_edit_patient\n",
      "    response = self.client.post(\"/edit_patient\", data=data, follow_redirects=True)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1167, in post\n",
      "    return self.open(*args, **kw)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\testing.py\", line 235, in open\n",
      "    response = super().open(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1116, in open\n",
      "    response_parts = self.run_wsgi_app(request.environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 988, in run_wsgi_app\n",
      "    rv = run_wsgi_app(self.application, environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1264, in run_wsgi_app\n",
      "    app_rv = app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1498, in __call__\n",
      "    return self.wsgi_app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1473, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 880, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 865, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3695147458.py\", line 337, in edit_patient\n",
      "    client = MongoClient('mongodb+srv://0602750:strokedataassignment@strokedata.2lpyv.mongodb.net/')\n",
      "Call MongoClient.close() to safely shut down your client and free up resources.\n",
      "  return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py:865: ResourceWarning: Unclosed MongoClient opened at:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 701, in start\n",
      "    self.io_loop.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\windows_events.py\", line 322, in run_forever\n",
      "    super().run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 641, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 1987, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 523, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 429, in dispatch_shell\n",
      "    await result\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 767, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 429, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3267539545.py\", line 261, in <module>\n",
      "    unittest.main(argv=[''], verbosity=2, exit=False)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 105, in __init__\n",
      "    self.runTests()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 281, in runTests\n",
      "    self.result = testRunner.run(self.test)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\runner.py\", line 240, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 690, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 634, in run\n",
      "    self._callTestMethod(testMethod)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 589, in _callTestMethod\n",
      "    if method() is not None:\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3267539545.py\", line 88, in test_edit_patient\n",
      "    response = self.client.post(\"/edit_patient\", data=data, follow_redirects=True)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1167, in post\n",
      "    return self.open(*args, **kw)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\testing.py\", line 235, in open\n",
      "    response = super().open(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1150, in open\n",
      "    response = self.resolve_redirect(response, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1056, in resolve_redirect\n",
      "    return self.open(builder, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\testing.py\", line 235, in open\n",
      "    response = super().open(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1116, in open\n",
      "    response_parts = self.run_wsgi_app(request.environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 988, in run_wsgi_app\n",
      "    rv = run_wsgi_app(self.application, environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1264, in run_wsgi_app\n",
      "    app_rv = app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1498, in __call__\n",
      "    return self.wsgi_app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1473, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 880, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 865, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3695147458.py\", line 211, in view_patients\n",
      "    client = MongoClient('mongodb+srv://0602750:strokedataassignment@strokedata.2lpyv.mongodb.net/')\n",
      "Call MongoClient.close() to safely shut down your client and free up resources.\n",
      "  return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "ok\n",
      "test_home (__main__.FlaskAppTests.test_home) ... ok\n",
      "test_logged_in (__main__.FlaskAppTests.test_logged_in) ... ok\n",
      "test_login_functionality (__main__.FlaskAppTests.test_login_functionality) ... ok\n",
      "test_login_page (__main__.FlaskAppTests.test_login_page) ... ok\n",
      "test_mongodb_data_functionality (__main__.FlaskAppTests.test_mongodb_data_functionality) ... C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py:865: ResourceWarning: Unclosed MongoClient opened at:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 701, in start\n",
      "    self.io_loop.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\windows_events.py\", line 322, in run_forever\n",
      "    super().run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 641, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 1987, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 523, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 429, in dispatch_shell\n",
      "    await result\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 767, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 429, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3267539545.py\", line 261, in <module>\n",
      "    unittest.main(argv=[''], verbosity=2, exit=False)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 105, in __init__\n",
      "    self.runTests()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 281, in runTests\n",
      "    self.result = testRunner.run(self.test)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\runner.py\", line 240, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 690, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 634, in run\n",
      "    self._callTestMethod(testMethod)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 589, in _callTestMethod\n",
      "    if method() is not None:\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3267539545.py\", line 108, in test_mongodb_data_functionality\n",
      "    response = self.client.get('/data')\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1162, in get\n",
      "    return self.open(*args, **kw)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\testing.py\", line 235, in open\n",
      "    response = super().open(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1116, in open\n",
      "    response_parts = self.run_wsgi_app(request.environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 988, in run_wsgi_app\n",
      "    rv = run_wsgi_app(self.application, environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1264, in run_wsgi_app\n",
      "    app_rv = app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1498, in __call__\n",
      "    return self.wsgi_app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1473, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 880, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 865, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3695147458.py\", line 403, in index\n",
      "    client = MongoClient('mongodb+srv://0602750:strokedataassignment@strokedata.2lpyv.mongodb.net/')\n",
      "Call MongoClient.close() to safely shut down your client and free up resources.\n",
      "  return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "ok\n",
      "test_mongodb_data_page (__main__.FlaskAppTests.test_mongodb_data_page) ... C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py:865: ResourceWarning: Unclosed MongoClient opened at:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 701, in start\n",
      "    self.io_loop.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\windows_events.py\", line 322, in run_forever\n",
      "    super().run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 641, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 1987, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 523, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 429, in dispatch_shell\n",
      "    await result\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 767, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 429, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3267539545.py\", line 261, in <module>\n",
      "    unittest.main(argv=[''], verbosity=2, exit=False)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 105, in __init__\n",
      "    self.runTests()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 281, in runTests\n",
      "    self.result = testRunner.run(self.test)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\runner.py\", line 240, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 690, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 634, in run\n",
      "    self._callTestMethod(testMethod)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 589, in _callTestMethod\n",
      "    if method() is not None:\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3267539545.py\", line 101, in test_mongodb_data_page\n",
      "    response = self.client.get('/data')\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1162, in get\n",
      "    return self.open(*args, **kw)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\testing.py\", line 235, in open\n",
      "    response = super().open(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1116, in open\n",
      "    response_parts = self.run_wsgi_app(request.environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 988, in run_wsgi_app\n",
      "    rv = run_wsgi_app(self.application, environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1264, in run_wsgi_app\n",
      "    app_rv = app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1498, in __call__\n",
      "    return self.wsgi_app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1473, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 880, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 865, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3695147458.py\", line 403, in index\n",
      "    client = MongoClient('mongodb+srv://0602750:strokedataassignment@strokedata.2lpyv.mongodb.net/')\n",
      "Call MongoClient.close() to safely shut down your client and free up resources.\n",
      "  return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "ok\n",
      "test_success (__main__.FlaskAppTests.test_success) ... ok\n",
      "test_upload_csv_file (__main__.FlaskAppTests.test_upload_csv_file) ... C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py:865: ResourceWarning: Unclosed MongoClient opened at:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 701, in start\n",
      "    self.io_loop.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\windows_events.py\", line 322, in run_forever\n",
      "    super().run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 641, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 1987, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 523, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 429, in dispatch_shell\n",
      "    await result\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 767, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 429, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3267539545.py\", line 261, in <module>\n",
      "    unittest.main(argv=[''], verbosity=2, exit=False)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 105, in __init__\n",
      "    self.runTests()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 281, in runTests\n",
      "    self.result = testRunner.run(self.test)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\runner.py\", line 240, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 690, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 634, in run\n",
      "    self._callTestMethod(testMethod)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 589, in _callTestMethod\n",
      "    if method() is not None:\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3267539545.py\", line 193, in test_upload_csv_file\n",
      "    response = client.post('/upload', content_type='multipart/form-data', data=data, follow_redirects=True)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1167, in post\n",
      "    return self.open(*args, **kw)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\testing.py\", line 235, in open\n",
      "    response = super().open(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1116, in open\n",
      "    response_parts = self.run_wsgi_app(request.environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 988, in run_wsgi_app\n",
      "    rv = run_wsgi_app(self.application, environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1264, in run_wsgi_app\n",
      "    app_rv = app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1498, in __call__\n",
      "    return self.wsgi_app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1473, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 880, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 865, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3695147458.py\", line 437, in upload\n",
      "    client = MongoClient('mongodb+srv://0602750:strokedataassignment@strokedata.2lpyv.mongodb.net/')\n",
      "Call MongoClient.close() to safely shut down your client and free up resources.\n",
      "  return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py:865: ResourceWarning: Unclosed MongoClient opened at:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 701, in start\n",
      "    self.io_loop.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\windows_events.py\", line 322, in run_forever\n",
      "    super().run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 641, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 1987, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 523, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 429, in dispatch_shell\n",
      "    await result\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 767, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 429, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3267539545.py\", line 261, in <module>\n",
      "    unittest.main(argv=[''], verbosity=2, exit=False)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 105, in __init__\n",
      "    self.runTests()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 281, in runTests\n",
      "    self.result = testRunner.run(self.test)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\runner.py\", line 240, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 690, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 634, in run\n",
      "    self._callTestMethod(testMethod)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 589, in _callTestMethod\n",
      "    if method() is not None:\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3267539545.py\", line 193, in test_upload_csv_file\n",
      "    response = client.post('/upload', content_type='multipart/form-data', data=data, follow_redirects=True)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1167, in post\n",
      "    return self.open(*args, **kw)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\testing.py\", line 235, in open\n",
      "    response = super().open(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1150, in open\n",
      "    response = self.resolve_redirect(response, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1056, in resolve_redirect\n",
      "    return self.open(builder, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\testing.py\", line 235, in open\n",
      "    response = super().open(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1116, in open\n",
      "    response_parts = self.run_wsgi_app(request.environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 988, in run_wsgi_app\n",
      "    rv = run_wsgi_app(self.application, environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1264, in run_wsgi_app\n",
      "    app_rv = app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1498, in __call__\n",
      "    return self.wsgi_app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1473, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 880, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 865, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3695147458.py\", line 211, in view_patients\n",
      "    client = MongoClient('mongodb+srv://0602750:strokedataassignment@strokedata.2lpyv.mongodb.net/')\n",
      "Call MongoClient.close() to safely shut down your client and free up resources.\n",
      "  return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py:589: ResourceWarning: Unclosed MongoClient opened at:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 701, in start\n",
      "    self.io_loop.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\windows_events.py\", line 322, in run_forever\n",
      "    super().run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 641, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 1987, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 523, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 429, in dispatch_shell\n",
      "    await result\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 767, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 429, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3267539545.py\", line 261, in <module>\n",
      "    unittest.main(argv=[''], verbosity=2, exit=False)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 105, in __init__\n",
      "    self.runTests()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 281, in runTests\n",
      "    self.result = testRunner.run(self.test)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\runner.py\", line 240, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 690, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 634, in run\n",
      "    self._callTestMethod(testMethod)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 589, in _callTestMethod\n",
      "    if method() is not None:\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3267539545.py\", line 198, in test_upload_csv_file\n",
      "    client = MongoClient('mongodb+srv://0602750:strokedataassignment@strokedata.2lpyv.mongodb.net/')\n",
      "Call MongoClient.close() to safely shut down your client and free up resources.\n",
      "  if method() is not None:\n",
      "ok\n",
      "test_upload_csv_page (__main__.FlaskAppTests.test_upload_csv_page) ... C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py:865: ResourceWarning: Unclosed MongoClient opened at:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 701, in start\n",
      "    self.io_loop.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\windows_events.py\", line 322, in run_forever\n",
      "    super().run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 641, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 1987, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 523, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 429, in dispatch_shell\n",
      "    await result\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 767, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 429, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3267539545.py\", line 261, in <module>\n",
      "    unittest.main(argv=[''], verbosity=2, exit=False)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 105, in __init__\n",
      "    self.runTests()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 281, in runTests\n",
      "    self.result = testRunner.run(self.test)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\runner.py\", line 240, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 690, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 634, in run\n",
      "    self._callTestMethod(testMethod)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 589, in _callTestMethod\n",
      "    if method() is not None:\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3267539545.py\", line 179, in test_upload_csv_page\n",
      "    response = self.client.get('/upload')\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1162, in get\n",
      "    return self.open(*args, **kw)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\testing.py\", line 235, in open\n",
      "    response = super().open(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1116, in open\n",
      "    response_parts = self.run_wsgi_app(request.environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 988, in run_wsgi_app\n",
      "    rv = run_wsgi_app(self.application, environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1264, in run_wsgi_app\n",
      "    app_rv = app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1498, in __call__\n",
      "    return self.wsgi_app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1473, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 880, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 865, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3695147458.py\", line 437, in upload\n",
      "    client = MongoClient('mongodb+srv://0602750:strokedataassignment@strokedata.2lpyv.mongodb.net/')\n",
      "Call MongoClient.close() to safely shut down your client and free up resources.\n",
      "  return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "ok\n",
      "test_user_delete_render (__main__.FlaskAppTests.test_user_delete_render) ... ok\n",
      "test_view_patients (__main__.FlaskAppTests.test_view_patients) ... "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "User data inserted into the database\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py:865: ResourceWarning: Unclosed MongoClient opened at:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 701, in start\n",
      "    self.io_loop.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\windows_events.py\", line 322, in run_forever\n",
      "    super().run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 641, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 1987, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 523, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 429, in dispatch_shell\n",
      "    await result\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 767, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 429, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3267539545.py\", line 261, in <module>\n",
      "    unittest.main(argv=[''], verbosity=2, exit=False)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 105, in __init__\n",
      "    self.runTests()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 281, in runTests\n",
      "    self.result = testRunner.run(self.test)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\runner.py\", line 240, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 690, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 634, in run\n",
      "    self._callTestMethod(testMethod)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 589, in _callTestMethod\n",
      "    if method() is not None:\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3267539545.py\", line 226, in test_view_patients\n",
      "    response = self.client.get('/view_patients')\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1162, in get\n",
      "    return self.open(*args, **kw)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\testing.py\", line 235, in open\n",
      "    response = super().open(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1116, in open\n",
      "    response_parts = self.run_wsgi_app(request.environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 988, in run_wsgi_app\n",
      "    rv = run_wsgi_app(self.application, environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1264, in run_wsgi_app\n",
      "    app_rv = app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1498, in __call__\n",
      "    return self.wsgi_app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1473, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 880, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 865, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3695147458.py\", line 211, in view_patients\n",
      "    client = MongoClient('mongodb+srv://0602750:strokedataassignment@strokedata.2lpyv.mongodb.net/')\n",
      "Call MongoClient.close() to safely shut down your client and free up resources.\n",
      "  return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "ok\n",
      "test_view_patients_search (__main__.FlaskAppTests.test_view_patients_search) ... C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py:865: ResourceWarning: Unclosed MongoClient opened at:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 701, in start\n",
      "    self.io_loop.start()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\windows_events.py\", line 322, in run_forever\n",
      "    super().run_forever()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 641, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\base_events.py\", line 1987, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\asyncio\\events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 523, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 429, in dispatch_shell\n",
      "    await result\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 767, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 429, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3267539545.py\", line 261, in <module>\n",
      "    unittest.main(argv=[''], verbosity=2, exit=False)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 105, in __init__\n",
      "    self.runTests()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\main.py\", line 281, in runTests\n",
      "    self.result = testRunner.run(self.test)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\runner.py\", line 240, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 84, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\suite.py\", line 122, in run\n",
      "    test(result)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 690, in __call__\n",
      "    return self.run(*args, **kwds)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 634, in run\n",
      "    self._callTestMethod(testMethod)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\unittest\\case.py\", line 589, in _callTestMethod\n",
      "    if method() is not None:\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3267539545.py\", line 251, in test_view_patients_search\n",
      "    response = self.client.get(f'/view_patients?search={search_query}')\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1162, in get\n",
      "    return self.open(*args, **kw)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\testing.py\", line 235, in open\n",
      "    response = super().open(\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1116, in open\n",
      "    response_parts = self.run_wsgi_app(request.environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 988, in run_wsgi_app\n",
      "    rv = run_wsgi_app(self.application, environ, buffered=buffered)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\werkzeug\\test.py\", line 1264, in run_wsgi_app\n",
      "    app_rv = app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1498, in __call__\n",
      "    return self.wsgi_app(environ, start_response)\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 1473, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 880, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\Users\\liann\\anaconda3\\Lib\\site-packages\\flask\\app.py\", line 865, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "  File \"C:\\Users\\liann\\AppData\\Local\\Temp\\ipykernel_33072\\3695147458.py\", line 211, in view_patients\n",
      "    client = MongoClient('mongodb+srv://0602750:strokedataassignment@strokedata.2lpyv.mongodb.net/')\n",
      "Call MongoClient.close() to safely shut down your client and free up resources.\n",
      "  return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]\n",
      "ok\n",
      "\n",
      "----------------------------------------------------------------------\n",
      "Ran 17 tests in 3.801s\n",
      "\n",
      "OK\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<unittest.main.TestProgram at 0x1798c21c320>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 4
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "6fcb929bb99598e8"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
